dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1083 @@
+    for (size_type v = 0; v < this->num_variables(); ++v) {
+        if (this->vartype(v) != other.vartype(v)) {
+            return false;
+        }
+    }
+
+    // check quadratic. We already checked the number of interactions so
+    // if one is present we can assume it's empty
+    if (this->has_adj() && other.has_adj()) {
+        // check that the neighborhoods are equal
+        auto it1 = this->cbegin_quadratic();
+        auto it2 = other.cbegin_quadratic();
+        for (; it1 != this->cend_quadratic(); ++it1, ++it2) {
+            if (*it1 != *it2) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+template <class bias_type, class index_type>
+bool QuadraticModelBase<bias_type, index_type>::is_linear() const {
+    if (has_adj()) {
+        for (const auto& n : *adj_ptr_) {
+            if (n.size()) return false;
+        }
+    }
+    return true;
+}
+
+template <class bias_type, class index_type>
+bias_type QuadraticModelBase<bias_type, index_type>::linear(index_type v) const {
+    assert(0 <= v && static_cast<size_t>(v) <= num_variables());
+    return linear_biases_[v];
+}
+
+template <class bias_type, class index_type>
+std::size_t QuadraticModelBase<bias_type, index_type>::nbytes(bool capacity) const {
+    size_type count = sizeof(bias_type);  // offset
+    if (capacity) {
+        count += linear_biases_.capacity() * sizeof(bias_type);
+    } else {
+        count += linear_biases_.size() * sizeof(bias_type);
+    }
+    if (has_adj()) {
+        if (capacity) {
+            for (const auto& n : (*adj_ptr_)) {
+                count += n.capacity() * sizeof(LinearTerm<bias_type, index_type>);
+            }
+        } else {
+            for (const auto& n : (*adj_ptr_)) {
+                count += n.size() * sizeof(LinearTerm<bias_type, index_type>);
+            }
+        }
+    }
+    return count;
+}
+
+template <class bias_type, class index_type>
+std::size_t QuadraticModelBase<bias_type, index_type>::num_interactions() const {
+    size_type count = 0;
+    if (has_adj()) {
+        index_type v = 0;
+        for (auto& n : *(adj_ptr_)) {
+            count += n.size();
+
+            // account for self-loops
+            auto lb = std::lower_bound(n.cbegin(), n.cend(), v);
+            if (lb != n.cend() && lb->v == v) {
+                count += 1;
+            }
+
+            ++v;
+        }
+    }
+    return count / 2;
+}
+
+template <class bias_type, class index_type>
+std::size_t QuadraticModelBase<bias_type, index_type>::num_interactions(index_type v) const {
+    if (has_adj()) {
+        return (*adj_ptr_)[v].size();
+    } else {
+        return 0;
+    }
+}
+
+template <class bias_type, class index_type>
+bias_type QuadraticModelBase<bias_type, index_type>::offset() const {
+    return offset_;
+}
+
+template <class bias_type, class index_type>
+bias_type QuadraticModelBase<bias_type, index_type>::quadratic(index_type u, index_type v) const {
```

Comment:
This one does not return a reference but the asym_quad does, so probably better name that one as asym_quadratic_ref ?

Reply:
Will rename to `asymmetric_quadratic_ref`